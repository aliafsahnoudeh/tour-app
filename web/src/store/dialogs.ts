import { defineStore } from "pinia";
import { ref } from "vue";

export const useDialogsStore = defineStore("dialogs", () => {
  const driverDialog = ref<boolean>(false);
  const deleteItemDialog = ref<boolean>(false);
  const deleteItemsDialog = ref<boolean>(false);

  const setDeleteItemsDialog = (input: boolean) => {
    deleteItemsDialog.value = input;
  };

  const setDeleteItemDialog = (input: boolean) => {
    deleteItemDialog.value = input;
  };

  const setDriverDialog = (input: boolean) => {
    driverDialog.value = input;
  };

  return {
    driverDialog,
    deleteItemDialog,
    deleteItemsDialog,
    setDeleteItemsDialog,
    setDriverDialog,
    setDeleteItemDialog,
  };
});
