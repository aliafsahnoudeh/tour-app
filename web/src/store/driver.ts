import { defineStore } from "pinia";
import { ref } from "vue";
import type DriverModel from "../types/DriverModel";

export const useDriverStore = defineStore("driver", () => {
  const driver = ref<DriverModel | undefined>(undefined);

  const setDriver = (payload: DriverModel | undefined) => {
    driver.value = payload;
  };

  return { driver, setDriver };
});
