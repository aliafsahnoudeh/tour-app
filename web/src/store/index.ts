import { defineStore } from 'pinia'


const useStore = defineStore('store', {
  state: () => {
    return { }
  },
  actions: {
  }
});

export default useStore;
