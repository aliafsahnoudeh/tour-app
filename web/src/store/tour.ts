import { defineStore } from "pinia";
import { ref } from "vue";
import type TourModel from "@/types/TourModel";

export const useTourStore = defineStore("tour", () => {
  const tour = ref<TourModel | undefined>(undefined);

  const setTour = (payload: TourModel | undefined) => {
    tour.value = payload;
  };

  return { tour, setTour };
});
