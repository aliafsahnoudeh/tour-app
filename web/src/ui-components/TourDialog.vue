<template>
  <div>
    <Dialog
      v-model:visible="dialogStore.driverDialog"
      :style="{ width: '450px' }"
      header="Product Details"
      :modal="true"
      class="p-fluid"
    >
      <div class="field">
        <label for="CustomerName">CustomerName</label>
        <InputText
          id="CustomerName"
          v-if="store.tour"
          v-model.trim="store.tour.CustomerName"
          required="true"
          autofocus
          :class="{ 'p-invalid': submitted && !store.tour?.CustomerName }"
        />
        <small class="p-error" v-if="submitted && !store.tour?.CustomerName"
          >CustomerName is required.</small
        >
      </div>
      <div class="field">
        <label for="ShipmentDate">ShipmentDate</label>
        <InputText
          id="ShipmentDate"
          v-if="store.tour"
          v-model.trim="store.tour.ShipmentDate"
          required="true"
          autofocus
          :class="{ 'p-invalid': submitted && !store.tour.ShipmentDate }"
        />
        <small class="p-error" v-if="submitted && !store.tour?.ShipmentDate"
          >ShipmentDate is required.</small
        >
      </div>
      <div class="field">
        <label for="LocationFrom">LocationFrom</label>
        <InputText
          id="LocationFrom"
          v-if="store.tour"
          v-model.trim="store.tour.LocationFrom"
          required="true"
          autofocus
          :class="{ 'p-invalid': submitted && !store.tour.LocationFrom }"
          @keypress="preventNumericInput"
        />
        <small class="p-error" v-if="submitted && !store.tour?.LocationFrom"
          >LocationFrom is required.</small
        >
      </div>
      <div class="field">
        <label for="LocationTo">LocationTo</label>
        <InputText
          id="LocationTo"
          v-if="store.tour"
          v-model.trim="store.tour.LocationTo"
          required="true"
          autofocus
          :class="{ 'p-invalid': submitted && !store.tour.LocationTo }"
          @keypress="preventNumericInput"
        />
        <small class="p-error" v-if="submitted && !store.tour?.LocationTo"
          >LocationTo is required.</small
        >
      </div>

      <Dropdown
        v-model="selectedDriver"
        :options="drivers"
        optionLabel="FirstName"
        placeholder="Select a Driver"
        @change="handleChange"
      />

      <template #footer>
        <Button
          label="Cancel"
          icon="pi pi-times"
          class="p-button-text"
          @click="emit('hide')"
        />
        <Button
          label="Save"
          icon="pi pi-check"
          class="p-button-text"
          @click="emit('save')"
        />
      </template>
    </Dialog>
  </div>
</template>

<script setup lang="ts">
import { defineEmits, watch, ref } from "vue";
import { useTourStore } from "../store/tour";
import { useDialogsStore } from "../store/dialogs";
import useDrivers from "../logical-components/useDrivers";
import type DriverModel from "../types/DriverModel";

const props = defineProps<{
  driverDialog: boolean;
  submitted: boolean;
}>();

const emit = defineEmits(["save", "hide"]);

const store = useTourStore();
const dialogStore = useDialogsStore();
const { drivers, fetchByLocation } = useDrivers(false);
const selectedDriver = ref<DriverModel>();

const preventNumericInput = (event: any) => {
  const keyCode = event.keyCode ? event.keyCode : event.which;
  if (keyCode > 47 && keyCode < 58) {
    event.preventDefault();
  }
};

watch(
  () => store.tour?.LocationFrom,
  (newValue) => {
    if (newValue !== undefined && newValue.length > 0)
      fetchByLocation(newValue);
  }
);

const handleChange = (event: any) => {
  selectedDriver.value = event.value;
  if (store.tour !== undefined) store.tour.Driver = event.value;
};
</script>
