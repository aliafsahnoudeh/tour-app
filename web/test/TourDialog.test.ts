import { describe, expect, test, vi, afterEach, beforeEach } from "vitest";
import { createPinia, setActivePinia } from "pinia";
import { mount, shallowMount } from "@vue/test-utils";
import { useDialogsStore } from "../src/store/dialogs";
import { useTourStore } from "../src/store/tour";
import type TourModel from "../src/types/TourModel";

import PrimeVue from "primevue/config";
import Card from "primevue/card";
import Button from "primevue/button";
import Avatar from "primevue/avatar";
import DataTable from "primevue/datatable";
import InputText from "primevue/inputtext";
import Toolbar from "primevue/toolbar";
import Column from "primevue/column";
import Dialog from "primevue/dialog";
import Dropdown from "primevue/dropdown";
import Calendar from "primevue/calendar";
import Menubar from "primevue/menubar";

import TourDialog from "../src/ui-components/TourDialog.vue";
import { nextTick } from "vue";

describe("TourDialog", () => {
  let wrapper;

  beforeEach(() => {
    setActivePinia(createPinia());

    wrapper = shallowMount(TourDialog, {
      propsData: {
        submitted: false,
      },
      global: {
        components: { Dialog, InputText, Calendar, Dropdown, Button },
        plugins: [PrimeVue],
        mocks: {
          ["$primevue"]: {
            config: {
              ripple: true,
            },
          },
        },
        provide: {
          services: vi.fn(),
        },
      },
    });
  });

  afterEach(() => {
    vi.clearAllMocks();
  });

  test("match the default snapshot", async () => {
    expect(TourDialog).toBeTruthy();

    expect(wrapper.element).toMatchSnapshot();
  });

  test("should shows fileds correctly", async () => {
    expect(TourDialog).toBeTruthy();
    const store = useDialogsStore();
    const tourStore = useTourStore();

    tourStore.setTour({
      Id: "",
      CustomerName: "test",
      LocationFrom: "a",
      LocationTo: "b",
      Driver: undefined,
      DriverId: undefined,
      ShipmentDate: new Date(),
    } as TourModel);

    store.setDriverDialog(true);

    await nextTick();

    expect(wrapper.element).toMatchSnapshot();
  });
});
