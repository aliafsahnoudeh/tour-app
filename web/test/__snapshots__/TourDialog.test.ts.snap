// Vitest Snapshot v1

exports[`TourDialog > match the default snapshot 1`] = `
<div>
  <dialog-stub
    appendto="body"
    autozindex="true"
    basezindex="0"
    class="p-fluid"
    closable="true"
    closeicon="pi pi-times"
    closeonescape="true"
    dismissablemask="false"
    draggable="true"
    header="Product Details"
    keepinviewport="true"
    maximizable="false"
    maximizeicon="pi pi-window-maximize"
    minimizeicon="pi pi-window-minimize"
    minx="0"
    miny="0"
    modal="true"
    position="center"
    showheader="true"
    style="width: 450px;"
    visible="false"
  />
</div>
`;

exports[`TourDialog > should shows fileds correctly 1`] = `
<div>
  <dialog-stub
    appendto="body"
    autozindex="true"
    basezindex="0"
    class="p-fluid"
    closable="true"
    closeicon="pi pi-times"
    closeonescape="true"
    dismissablemask="false"
    draggable="true"
    header="Product Details"
    keepinviewport="true"
    maximizable="false"
    maximizeicon="pi pi-window-maximize"
    minimizeicon="pi pi-window-minimize"
    minx="0"
    miny="0"
    modal="true"
    position="center"
    showheader="true"
    style="width: 450px;"
    visible="true"
  />
</div>
`;
